### ISSUE TITLE
Heading levels

### ISSUE THREAD

**Issue by dhardy** [raphlinus/pulldown-cmark#501]
The `Tag` enum includes:
```
    /// A heading. The field indicates the level of the heading.
    Heading(u32),
```
HTML and Markdown specs are both explicit that heading levels 1-6 are valid (in HTML called `h1`..`h6`). From this pulldown-cmark spec. it is not clear whether other values may occur (and thus how consumers should handle unexpected values).

Suggestion:
```rust
#[repr(u32)]
enum Heading {
    H1 = 1,
    H2,
    // ...
}
```
This supports `Heading::H6 as u32` and thus is just a small extension over the current status. Alternatively documentation should clarify exactly which values may appear.

**Comment by raphlinus** [comment]
I'm a little skeptical of this, as [*no* markdown implementations](https://babelmark.github.io/?text=%23%23%23%23%23%23%23+heading+7%3F) actually do H7.

**Comment by dhardy** [comment]
"Blogdown" does, but that's besides the point: I expect the API to document (via an enum or just with words) which values it is possible for the parser to generate.

**Comment by raphlinus** [comment]
Ah, I think I misunderstood, apologies. I can see the argument that an enum is more Rust-idiomatic, but it would be a breaking change. I certainly agree that documenting the range is a good idea (it is documented in the spec, but that's not as accessible).

So my feeling is that a doc change would be quite welcome, and I'll let other people chime on whether the change to the enum is enough of an improvement to justify the (admittedly minor) breakage.

**Comment by edward-shen** [comment]
TL;DR I think this change is good as it promotes better semantic meaning, robustness, and better library behavior.

Going to chime in and say that I'm supporting of this (minor) breaking change and this should definitely be something to consider before a 1.0.0 release. I don't think there's any intention from WHATWG to add a 7th heading level nor is such an idea popular with any markdown implementation.

This kinda makes the contract of `Heading(u32)` a little awkward, as it implies that the client is expected to handle nonstandard levels that I think we can reasonably agree won't exist in the foreseeable future. It's like putting a `non_exhaustive` on a Heading enum -- sure, we can't guarantee that they won't add a new header level, but it's a little extreme to require every client to handle those (very) unlikely cases (I would assume most would do something like `_ -> unreachable!()` or just error out).

Restricting the values to only standard ones also follows the [robust principle](https://en.wikipedia.org/wiki/Robustness_principle):  Be conservative in what you send, be liberal in what you accept. From this perspective I'd even go on a limb to say we should implement all conversions to unsigned int types as well.

Finally, it allows for stricter compile time checking, as the compiler can guarantee that the value ranges are from H1-H6, and not the entire u32 range.

**Comment by marcusklaas** [comment]
This sounds like a sensible change. We welcome PRs to implement this.
